rwengine/src/audio/SoundManager.cpp

```diff
@@ -173,6 +178,226 @@ void SoundManager::SoundSource::loadFromFile(const rwfs::path& filePath) {
     avformat_close_input(&formatContext);
 }
 
+/// Structure for input data
+struct InputData {
+    uint8_t *ptr = nullptr;
+    size_t size{}; ///< size left in the buffer
+};
+
+/// Low level function for copying data from handler (opaque)
+/// to buffer.
+static int read_packet(void *opaque, uint8_t *buf, int buf_size) {
+    auto* input = reinterpret_cast<InputData*>(opaque);
+    buf_size = FFMIN(buf_size, input->size);
+    /* copy internal data to buf */
+    memcpy(buf, input->ptr, buf_size);
+    input->ptr  += buf_size;
+    input->size -= buf_size;
+    return buf_size;
+}
+
+void SoundManager::SoundSource::loadSfx(const rwfs::path& path, const size_t& index, const bool asWave) {
+    // Allocate audio frame
+    AVFrame* frame = av_frame_alloc();
+    if (!frame) {
+        RW_ERROR("Error allocating the audio frame");
+        return;
+    }
+
+    /// Now we need to prepare "custom" format context
+    /// We need sdt loader for that purpose
+    LoaderSDT sdt{};
+    sdt.load(path / "audio/sfx");
+
+    std::unique_ptr<char[]> raw_sound = sdt.loadToMemory(index, asWave);
+    if (!raw_sound) {
+        av_frame_free(&frame);
+        RW_ERROR("Error loading sound");
+        return;
+    }
+
+    /// Prepare input
+    InputData input{};
+    input.size = sizeof(WaveHeader) + sdt.assetInfo.size;
+    auto inputDataStart = std::make_unique<uint8_t[]>(input.size); /// Store start ptr of data to be able freed memory later
+    input.ptr = inputDataStart.get();
+
+    /// Alocate memory for buffer
+    /// Memory freeded at the end
+    static constexpr size_t ioBufferSize = 4096;
+    auto  ioBuffer = static_cast<uint8_t*>(av_malloc(ioBufferSize));
+
+    /// Cast pointer, in order to match required layout for ffmpeg
+    input.ptr = reinterpret_cast<uint8_t*>(raw_sound.get());
+
+    /// Finally prepare our "custom" format context
+    AVIOContext* avioContext = avio_alloc_context(ioBuffer, ioBufferSize, 0, &input, &read_packet, nullptr, nullptr);
+    AVFormatContext* formatContext = avformat_alloc_context();
+    formatContext->pb = avioContext;
+
+    if (avformat_open_input(&formatContext, "nothint", nullptr, nullptr) != 0) {
+        av_frame_free(&frame);
+        RW_ERROR("Error opening audio file (" << index << ")");
+        return;
```

Comment:
Should ``formatContext->pb`` be freed in these early branches?

Reply:
should be

Reply:
Done.